base/array.jl

```diff
@@ -627,6 +630,7 @@ oneunit(x::AbstractMatrix{T}) where {T} = _one(oneunit(T), x)
 ## Conversions ##
 
 convert(::Type{T}, a::AbstractArray) where {T<:Array} = a isa T ? a : T(a)
+convert(::Type{T}, a::AbstractArray) where {T<:Core.ImmutableArray} = a isa T ? a : Core.ImmutableArray(a)
```

Comment:
This doesn't ensure the return of something of type `T`. Perhaps it could be folded into a Union, but may be better to keep it separate, but with identical definitions?
```suggestion
convert(T::Union{Type{<:Array},Type{<:Core.ImmutableArray}}, a::AbstractArray) = a isa T ? a : T(a)
```
```suggestion
convert(::Type{T}, a::AbstractArray) where {T<:Array} = a isa T ? a : T(a)
convert(::Type{T}, a::AbstractArray) where {T<:Core.ImmutableArray} = a isa T ? a : T(a)
```

Reply:
Will fix this, just need to define some new constructors